core/trino-spi/src/main/java/io/trino/spi/connector/ConnectorMetadata.java

```diff
@@ -987,6 +987,25 @@ default ConnectorTableProperties getTableProperties(ConnectorSession session, Co
         return Optional.empty();
     }
 
+    /**
+     * Attempt to push down the join operation.
+     * <p>
+     * Connectors can indicate whether they don't support join pushdown or that the action had no effect
+     * by returning {@link Optional#empty()}. Connectors should expect this method may be called multiple times.
+     * </p>
+     */
+    default Optional<ConnectorTableHandle> applyJoin(
+            ConnectorSession session,
+            JoinType joinType,
+            ConnectorTableHandle left,
+            ConnectorTableHandle right,
+            ConnectorExpression joinCondition,
```

Comment:
Are you planning to extends `ConnectorExpression` so we actually can express join condition in separate PR?

Reply:
Good question.

There are two ways to handle this

- extend `ConnectorExpression` to accomodate this -- this is my _preferred_ way, and i can add this to this PR too
- expose equi join conditions in a dedicated manner and delay `ConnectorExpression` changes until later

I think we should be able to get some pre-agreement on which way to go without writing some boilerplate yet.

cc @martint @kokosing

Reply:
_preffered way_ will have influence on projection pushdown. I also think this is preferred way.

Reply:
> extend ConnectorExpression to accomodate this -- this is my preferred way, and i can add this to this PR too

This first needs the work to support representing function invocations in ConnectorExpression, which depends on the ongoing revamp @dain has been working on that will allow the engine to communicate the identity of functions to connectors and for connectors to query and look up functions.

Reply:
> This first needs the work to support representing function invocations in ConnectorExpression

Why equi predicates depends on function? Is it about some magic function and complex types? Maybe we can start with simple types.

Reply:
@martint i am aware of the function revamp, although i am not aware of the progress or  ETA. For this reason, i assume anything function-related _might_ be a no-go, so let's focus on what's possible today (eg equi conditions seem like a good candidate), keeping an eye on future compatibility when functions can be weaved in.

Reply:
chatted shortly with @martint about this and he suggested with explicit modelling equi clauses, as a beta/experimental/interim solution, to be replaced later with `ConnectorExpression`.
The reason is he envisions equality and all other operators will be represented in a new IR as function calls and will be represented as function calls in connector expressions too.

i've updated the PR to replace joinCondition with equi clauses.

Reply:
FYI, i've added inequalities too, as i didn't see a downside to it, especially given we consider this API not to be permanent and we replace it in the future with better ConnectorExpressions.
@martint lemme know if i am missing anything in my thinking